FpsCheck.py

```diff
@@ -366,33 +364,48 @@ def find_invalid_well_known(self, check_sets):
             try:
                 json_schema = self.open_and_load_json(url)
                 curr_fps_set = check_sets[primary]
-                schema_fields = set({field for field in json_schema.keys()
-                                     if field in self.acceptable_fields})
-                schema_fields.update({field for field in self.acceptable_fields
-                        if len(curr_fps_set.relevant_fields_dict[field]) > 0})
-                for field in schema_fields:
-                    if field == "primary":
-                        if json_schema["primary"] != curr_fps_set.primary:
-                            field_sym_difference = [json_schema["primary"], 
-                            curr_fps_set.primary]
-                        else:
-                            field_sym_difference = []
-                    else:
-                        field_sym_difference = set(json_schema[field] 
-                                if field in json_schema.keys() else []) ^ set(
-                                    curr_fps_set.relevant_fields_dict[field])
-                        if field == 'ccTLDs':
-                            for aliased_site in json_schema[field]:
-                                field_sym_difference.update(
-                                    set(json_schema[field][aliased_site]) ^ 
-                                    set(
-                                    curr_fps_set.relevant_fields_dict[field]
-                                    [aliased_site]))
-                    if field_sym_difference:
-                        self.error_list.append("The following member(s) of " 
-                        + field + " were not present in both the changelist "
-                        + "and .well-known/first-party-set.json file: " + 
-                        str(sorted(field_sym_difference)))
+                well_known_set = FpsSet(
+                    json_schema.get('ccTLDs'), 
+                    json_schema.get('primary'), 
+                    json_schema.get('associatedSites'), 
+                    json_schema.get('serviceSites'))
+                if well_known_set.primary != curr_fps_set.primary:
+                    field_sym_difference = [well_known_set.primary, curr_fps_set.primary]
+                    self.error_list.append("The following primary was not "
+                    + "present in both the changelist and "
+                    + ".well-known/first-party-set.json file: "
+                    + str(sorted(field_sym_difference)))
+                if well_known_set.associated_sites != curr_fps_set.associated_sites:
+                    field_sym_difference = set(
+                       well_known_set.associated_sites) ^ set(curr_fps_set.associated_sites)
```

Comment:
The fact that we're checking equality of the _lists_, but only displaying the symmetric difference between the _sets_ means that comparing `['a', 'b']` and `['b', 'a']` would give an error like `"The following member(s) of associatedSites were not present in both the changelist and .well-known/first-party-set.json file: []"`. That seems like it would be pretty hard to debug.

Ideally, we would print just the necessary and sufficient information needed to fix the issue in the PR (possibly by computing the Levenshtein distance and the corresponding edits between the two lists?). But in the absence of that, I think it would be better to print the two input lists themselves, and also print the difference (either as a symmetric difference, or as both asymmetric differences). Then the error message would contain enough to diagnose what the issue is, without digging into the repo's code - even though it's more verbose. WDYT?

FWIW, I do think it's good to be sensitive to the order that the sites are listed in the subsets, since Chrome has a limit for the number of associated sites it accepts into a set. Enforcing that the .well-known files declare sites in the same order as the GitHub repo does will help ensure consistency of the ordering between the different set declarations.

Reply:
Right, I see the issue there. I agree with the thought of printing out the lists to better communicate the issue to the submitter. However, would we just want to do this for the associatedSites field since that is the only one in which the order will have an impact? Or should we do it for the other fields as well?

Reply:
I would enforce the ordering for all subsets (+ ccTLDs). I don't see a reason not to, and it's much easier to be strict *now* than to discover we need to be strict later and try to migrate.